gentypes_test.go

```diff
@@ -0,0 +1,255 @@
+	{pb.Type_BYTES, "[]byte"},
+	{pb.Type_DATE, "time.Time"},
+	{pb.Type_DATETIME, "time.Time"},
+	{pb.Type_STRING, "string"},
+}
+var primitiveErrTests = []pb.Type_Primitive{
+	pb.Type_XML,
+	pb.Type_UUID,
+}
+
+func TestGenPrimitiveType(tt *testing.T) {
+	assert := testifyAssert.New(tt)
+	for _, t := range primitiveTests {
+		str, err := GenPrimitiveType(&t.input)
+		assert.NoError(err)
+		assert.Equal(t.expected, str)
+	}
+	for _, tp := range primitiveErrTests {
+		_, err := GenPrimitiveType(&tp)
+		assert.Error(err)
+	}
+}
+
+func TestGenSimpleType(tt *testing.T) {
+	assert := testifyAssert.New(tt)
+	for _, t := range primitiveTests {
+		primitive := &pb.Type_Primitive_{Primitive: t.input}
+		pt := &pb.Type{Type: primitive}
+		str, err := GenSimpleType(pt)
+		assert.NoError(err)
+		assert.Equal(t.expected, str)
+	}
+	for _, tp := range primitiveErrTests {
+		primitive := &pb.Type_Primitive_{Primitive: tp}
+		pt := &pb.Type{Type: primitive}
+		_, err := GenSimpleType(pt)
+		assert.Error(err)
+	}
+
+	// test type refs
+	var refTests = []struct {
+		input    string
+		expected string
+	}{
+		{"MyType", "MyType"},
+		{"x", "x"},
+		{"map of string:int", "map[string]int"},
+	}
+	for _, t := range refTests {
+		ref := &pb.Scope{Path: []string{t.input}}
+		typeRef := &pb.Type_TypeRef{TypeRef: &pb.ScopedRef{Ref: ref}}
+		pt := &pb.Type{Type: typeRef}
+		str, err := GenSimpleType(pt)
+		assert.NoError(err)
+		assert.Equal(t.expected, str)
+	}
+	var errRefTests = []struct {
+		input []string
+	}{
+		{[]string{"MyType", "MyType"}},
+		{[]string{"map of string:int:bool"}},
+	}
+	for _, t := range errRefTests {
+		ref := &pb.Scope{Path: t.input}
+		typeRef := &pb.Type_TypeRef{TypeRef: &pb.ScopedRef{Ref: ref}}
+		pt := &pb.Type{Type: typeRef}
+		_, err := GenSimpleType(pt)
+		assert.Error(err)
+	}
+
+	strType := &pb.Type_Primitive_{Primitive: pb.Type_STRING}
+	list := &pb.Type_List_{List: &pb.Type_List{Type: &pb.Type{Type: strType}}}
+	pt := &pb.Type{Type: list}
+	_, err := GenSimpleType(pt)
+	assert.Error(err)
+}
+
+func TestSplitUppercase(tt *testing.T) {
+	assert := testifyAssert.New(tt)
+
+	assert.Equal([]string{"a"}, SplitUppercase("a"))
+	assert.Equal([]string{"abc"}, SplitUppercase("abc"))
+	assert.Equal([]string{"abc", "xyz"}, SplitUppercase("AbcXyz"))
+	assert.Equal([]string{"a", "b", "c"}, SplitUppercase("ABC"))
+	assert.Equal([]string{"a", "b", "c"}, SplitUppercase("aBC"))
+}
+
+func TestGenStruct(tt *testing.T) {
+	assert := testifyAssert.New(tt)
+
+	attrDefs := map[string]*pb.Type{}
+	typeTuple := &pb.Type_Tuple_{Tuple: &pb.Type_Tuple{AttrDefs: attrDefs}}
+	ttype := &pb.Type{Type: typeTuple}
+
+	attrDefs["Data"] = &pb.Type{
+		Type:          &pb.Type_Primitive_{Primitive: pb.Type_ANY},
+		SourceContext: &pb.SourceContext{Start: &pb.SourceContext_Location{Line: 1}},
+	}
+
+	str, err := GenStruct("DataPayload", ttype, "")
+	assert.NoError(err)
+	expectedSrc := `type DataPayload struct {
+			Data interface{} ` + "`" + `json:"Data"` + "`" + `
```

Comment:
This backquoting is madness! I can't think of a better option myself, though, so 🤷‍♂️.

Reply:
I know. It's either
```go
Data interface{} ` + "`" + `json:"Data"` + "`" + `
```
or
```go
Data interface{} ` + "`json:\"Data\"`" + `
```
both not beautiful.

Reply:
Maybe the second one is better....

Reply:
What do you think?

Reply:
Now that I see them next to each other I've changed it all to the second one.